### DISCUSSION TITLE
Make the failure as Lambda result with HTTP status code 4xx, 5xx

### DISCUSSION THREAD

**Discussion by mirumirumi** [aws/aws-lambda-rust-runtime#643] [Q&A]
I'm creating an HTTP response using the following code:

```rust
use lambda_http::{Body, Error, Response};

fn base_response_builder() -> http::response::Builder {
    Response::builder().header("Content-Type", "application/json")
}

pub fn ok(body: impl Into<Body>) -> Result<Response<Body>, Error> {
    Ok(base_response_builder().status(200).body(body.into())?)
}
```

I've also created this for 4xx and 5xx status codes, but since it ultimately returns an Ok from Lambda, the execution result of Lambda is considered "successful" rather than "failed". I would like the Lambda result to be treated as a failure when it's not a 2xx status code, so I want to return an Err at the end while setting the status code.

(My understanding is that if the Lambda Function returns Ok(), it is considered "success" and if it returns Err(), it is considered "failure", is that correct?)

How can I achieve this scenario?

**Comment by greenwoodcm** [comment]
You likely want to leave the function's return value as `Ok(_)`.  When you return an `Err(_)` to the runtime it will signal a failure back to the Lambda service, which tells the service that the execution environment has crashed or is otherwise bad.  That means that Lambda will restart the execution environment before issuing another invoke.  In most cases like you're describing, the code ran successfully, it just wants to signal an error from a downstream dependency or a logical error or something to the caller.  In such cases, you likely want the app to keep running and keep responding to subsequent requests - it was the request that failed, not the whole app.

So you should consider whether the entire application has now failed (thus return an `Err`), or if the single request failed and the app should keep running (thus return an `Ok`).  Does this help?

**Reply by mirumirumi** [reply]
@greenwoodcm 
I see.
I understand that it is a very bad idea to treat Lambda itself as a "failure", at least in 4xx series errors.

For actual server-related failures that occur, such as AWS SDK errors, "failure" might be fine, but even if it doesn't explicitly return a status code, API GW will return 502 if the execution result is a failure (which can be mapped to 500), and it's already good enough to be I noticed that.

Thanks!
